In [43]:
import sys , os
sys.path.append(os.pardir)
import numpy as np
from dataset.mnist import load_mnist

In [44]:
(x_train, t_train), (x_test, t_test) = \
    load_mnist(normalize=True, one_hot_label=True)

normalize 는 입력 이미지의 픽셀값을 0.0 ~ 1.0 사이의 값으로 정규화 할지 정함 / False니깐 0~255 를 유지하기 , one_hot_label 은 레이블을 형태로 저장할지 결정함 ex): [0,0,0,1,0,0,0] 처럼 정답을 뜻하는건 1(hot)이고 나머지 0은 one 0인 배열

In [45]:
print(x_train.shape)
print(t_train.shape)

(60000, 784)
(60000, 10)


위 출력을 보고 알수 있는것은
x_train은 60000개의 데이터가 훈련데이터고 , 784개가 입력 데이터열 28x28 이미지의 데이터임
t_train은 60000개의 데이터가 훈련데이터이고 , 10줄의 정답레이블

In [46]:
train_size = x_train.shape[0]
batch_size = 10
batch_mask = np.random.choice(train_size, batch_size)

x_batch = x_train[batch_mask]
t_train = t_train[batch_mask]

np.random.choice(60000, 10)

array([55580, 37505, 39843, 16195, 40789, 42582,  1369,  1541, 28527,
       48425])

In [48]:
def cross_entropy_error(y,t):
    if y.ndim == 1:
        t = t.reshape(1, t.size)
        y = y.reshape(1, t.size)

        batch_size = y[0]
        return -np.sum(t * np.log(y + 1e-7)) / batch_size

위 코드는 배치 데이터를 지원하는 교차 엔트로피 오차를 구현해본 코드
코드 설명 : y는 신경망의 출력 , t는 정답 레이블 만약 y가 1차원이라면 데이터 하나당 교차 엔트로피 오차를 구하는 경우 reshape 함수로 데이터의 형상을 변경해준다.
그리고 배치의 크기를 정규화 하고 , 이미지 1장당 평균 교차 엔트로피를 계산한다.